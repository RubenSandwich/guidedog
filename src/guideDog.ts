import { parseFragment, DefaultTreeElement, DefaultTreeTextNode } from 'parse5';

interface IGuideDogOptions {
  filterType?: GuideDogFilter;
  sourceCodeLoc?: boolean;
}

export enum GuideDogFilter {
  Headers,
}

type AccessibleNodeWithSource = AccessibleNode & { sourceCodeLoc: Location };
export type AccessibleNodes = AccessibleNode[] | AccessibleNodeWithSource[];

export const guideDog = (
  html: string,
  options?: IGuideDogOptions,
): AccessibleNodes => {
  const defaults: IGuideDogOptions = {
    filterType: GuideDogFilter.Headers,
    sourceCodeLoc: false,
  };

  const optionsWithDefaults = { ...defaults, ...options };

  const document = parseFragment(html, {
    sourceCodeLocationInfo: optionsWithDefaults.sourceCodeLoc,
  }) as DefaultTreeElement;

  const tree = parseIntoAccessibleNodes(document, optionsWithDefaults);

  return tree;

  // TODO: How do I get the type limited here?
  // return optionsWithDefaults.sourceCodeLoc
  //   ? (tree as AccessibleNodeWithSource[])
  //   : (tree as AccessibleNode[]);
};

const getFirstChild = (node: DefaultTreeElement) => {
  return node.childNodes[0];
};

const parseIntoAccessibleNodes = (
  node: DefaultTreeElement,
  options: IGuideDogOptions,
  accessibleNodes: AccessibleNodes = [],
) => {
  const { filterType, sourceCodeLoc } = options;

  if (!node.childNodes) {
    return accessibleNodes;
  }

  if (isHeading(node.tagName)) {
    const level = getHeadingLevel(node.tagName);
    const textNode = getFirstChild(node) as DefaultTreeTextNode;
    const insertIndex = getHeaderInsertIndex(accessibleNodes, level);

    const newNode: AccessibleNode = {
      role: 'heading',
      name: textNode.value,
      level,
      focusable: false,
    };

    if (sourceCodeLoc) {
      newNode.sourceCodeLoc = {
        startOffset: node.sourceCodeLocation.startOffset,
        endOffset: node.sourceCodeLocation.endOffset,
      };
    }

    return upsertNode(accessibleNodes, newNode, insertIndex);
  }

  let newAccessibleNodes = accessibleNodes;

  node.childNodes.forEach(childNode => {
    newAccessibleNodes = parseIntoAccessibleNodes(
      childNode as DefaultTreeElement,
      options,
      newAccessibleNodes,
    );
  });

  return newAccessibleNodes;
};

const isHeading = (nodeTagName: string): boolean => {
  return /^h[1-6]/.test(nodeTagName);
};

const getHeadingLevel = (nodeTagName: string): number => {
  return parseInt(nodeTagName.match(/[1-6]/)[0], 10);
};

export const getHeaderInsertIndex = (
  accessibleNodes: AccessibleNodes,
  insertHeaderLevel: number,
): number[] => {
  if (accessibleNodes.length === 0) {
    return [0];
  }

  const lastNodeIndex = accessibleNodes.length - 1;
  const lastNode = accessibleNodes[accessibleNodes.length - 1];

  if (lastNode.level >= insertHeaderLevel) {
    return [lastNodeIndex + 1];
  } else if (lastNode.children == null) {
    return [lastNodeIndex, 0];
  }

  return [
    lastNodeIndex,
    ...getHeaderInsertIndex(lastNode.children, insertHeaderLevel),
  ];
};

export const upsertNode = (
  accessibleNodes: AccessibleNodes,
  node: AccessibleNode,
  indexPath: number[],
): AccessibleNodes => {
  if (accessibleNodes.length == 0) {
    return [node];
  }

  const insertIndex = indexPath[0];

  if (indexPath.length === 1) {
    return [
      ...accessibleNodes.slice(0, insertIndex),
      node,
      ...accessibleNodes.slice(insertIndex),
    ];
  }

  const currentNode = accessibleNodes[insertIndex];

  return [
    ...accessibleNodes.slice(0, insertIndex),
    {
      ...currentNode,
      children: upsertNode(
        currentNode.children || [],
        node,
        indexPath.slice(1),
      ),
    },
    ...accessibleNodes.slice(insertIndex + 1),
  ];
};

interface SourceLocation {
  startOffset: number;
  endOffset: number;
}

interface AccessibleNode {
  role: string;
  name: string;

  // Only affects accessible focus
  focusable: boolean;
  level?: number;

  // Tree walking
  parent?: AccessibleNode;
  children?: AccessibleNode[];

  // Maybe?
  firstChild?: AccessibleNode;
  lastChild?: AccessibleNode;
  previousSibling?: AccessibleNode;
  nextSibling?: AccessibleNode;

  sourceCodeLoc?: SourceLocation;
}
