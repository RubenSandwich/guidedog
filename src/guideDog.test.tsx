import { guideDog, GuideDogFilter } from "./guideDog";

import * as React from "react";

const TestComp = () => {
  return <h1>header 1</h1>;
};

const TestComp2 = () => {
  return <h2>header 1, level 2</h2>;
};

const TestCompWithProps = (props: { text: string }) => {
  return <span>props.text</span>;
};

const TestCompLandmarks = () => {
  return (
    <body>
      <header>
        <div>raw text</div>
      </header>
      <main>
        <h1>header 1</h1>
      </main>
      <footer>
        <h1>header 2</h1>
      </footer>
    </body>
  );
};

const TestCompTabable = () => {
  return (
    <body>
      <header>
        <div>raw text</div>
      </header>
      <main>
        <h1>header 1</h1>
      </main>
      <footer>
        <a href="test">Link 1</a>
      </footer>
      <a href="test">Link 2</a>
      <div>
        <label htmlFor="name">Your Name:</label>
        <input name="name" type="text" />
      </div>
    </body>
  );
};

test("guideDog basic react comp", async () => {
  const accessibilityTree = await guideDog(<TestComp />);

  expect(accessibilityTree).toMatchSnapshot();
});

test("guideDog basic react comp 2", async () => {
  const accessibilityTree = await guideDog(<TestComp2 />);

  expect(accessibilityTree).toMatchSnapshot();
});

test("guideDog string", async () => {
  const accessibilityTree = await guideDog("<h1>header 1</h1>");

  expect(accessibilityTree).toMatchSnapshot();
});

test("guideDog string 2", async () => {
  const accessibilityTree = await guideDog(
    `<body>
      <header>
        <div>raw text</div>
      </header>
      <main>
        <h1>header 1</h1>
      </main>
      <footer>
        <h1>header 2</h1>
      </footer>
    </body>`
  );

  expect(accessibilityTree).toMatchSnapshot();
});

test("guideDog comp with props", async () => {
  const accessibilityTree = await guideDog(<TestCompWithProps text="great" />);

  expect(accessibilityTree).toMatchSnapshot();
});

test("guideDog landmarks", async () => {
  const accessibilityTree = await guideDog(<TestCompLandmarks />, {
    filterType: GuideDogFilter.OnlyLandmarks
  });

  expect(accessibilityTree).toMatchSnapshot();
});

test("guideDog tabable elements", async () => {
  const accessibilityTree = await guideDog(<TestCompTabable />);

  expect(accessibilityTree).toMatchSnapshot();

  const tabableTree = await guideDog(<TestCompTabable />, {
    filterType: GuideDogFilter.OnlyTabableElements
  });

  expect(tabableTree).toMatchSnapshot();
  expect(tabableTree).not.toBe(accessibilityTree);
});
