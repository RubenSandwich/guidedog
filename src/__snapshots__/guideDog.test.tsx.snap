// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`guideDog basic react comp 1`] = `
Array [
  Object {
    "html": "<h1 data-reactroot=\\"\\">header 1</h1>",
    "level": 1,
    "name": "header 1",
    "role": "heading",
  },
]
`;

exports[`guideDog basic react comp 2 1`] = `
Array [
  Object {
    "html": "<h2 data-reactroot=\\"\\">header 1, level 2</h2>",
    "level": 2,
    "name": "header 1, level 2",
    "role": "heading",
  },
]
`;

exports[`guideDog comp with props 1`] = `
Array [
  Object {
    "html": "props.text",
    "name": "props.text",
    "role": "text",
  },
]
`;

exports[`guideDog landmarks 1`] = `
Array [
  Object {
    "html": "<header><div>raw text</div></header>",
    "name": "",
    "role": "banner",
  },
  Object {
    "html": "<main><h1>header 1</h1></main>",
    "name": "",
    "role": "main",
  },
  Object {
    "html": "<footer><h1>header 2</h1></footer>",
    "name": "",
    "role": "contentinfo",
  },
]
`;

exports[`guideDog string 1`] = `
Array [
  Object {
    "html": "<h1>header 1</h1>",
    "level": 1,
    "name": "header 1",
    "role": "heading",
  },
]
`;

exports[`guideDog string 2 1`] = `
Array [
  Object {
    "html": "raw text",
    "name": "raw text",
    "role": "text",
  },
  Object {
    "html": "<h1>header 1</h1>",
    "level": 1,
    "name": "header 1",
    "role": "heading",
  },
  Object {
    "html": "<h1>header 2</h1>",
    "level": 1,
    "name": "header 2",
    "role": "heading",
  },
]
`;

exports[`guideDog tabable elements 1`] = `
Array [
  Object {
    "html": "raw text",
    "name": "raw text",
    "role": "text",
  },
  Object {
    "html": "<h1>header 1</h1>",
    "level": 1,
    "name": "header 1",
    "role": "heading",
  },
  Object {
    "html": "<a href=\\"test\\">Link 1</a>",
    "name": "Link 1",
    "role": "link",
  },
  Object {
    "html": "<a href=\\"test\\">Link 2</a>",
    "name": "Link 2",
    "role": "link",
  },
  Object {
    "html": "Your Name:",
    "name": "Your Name:",
    "role": "text",
  },
  Object {
    "html": "<input type=\\"text\\" name=\\"name\\">",
    "name": "",
    "role": "textbox",
  },
]
`;

exports[`guideDog tabable elements 2`] = `
Array [
  Object {
    "html": "<a href=\\"test\\">Link 1</a>",
    "name": "Link 1",
    "role": "link",
  },
  Object {
    "html": "<a href=\\"test\\">Link 2</a>",
    "name": "Link 2",
    "role": "link",
  },
  Object {
    "html": "<input type=\\"text\\" name=\\"name\\">",
    "name": "",
    "role": "textbox",
  },
]
`;
